prov/rxm/src/rxm_ep.c

```diff
@@ -662,117 +800,106 @@ rxm_ep_send_common(struct fid_ep *ep_fid, const struct iovec *iov, void **desc,
-	}
-
-	if ((flags & FI_INJECT) && !(flags & FI_COMPLETION)) {
-		if (pkt_size <= rxm_ep->msg_info->tx_attr->inject_size) {
-			if (tx_entry->state == RXM_LMT_TX) {
-				RXM_LOG_STATE_TX(FI_LOG_EP_DATA, tx_entry,
-						 RXM_LMT_TX);
-				tx_entry->state = RXM_LMT_ACK_WAIT;
+		if ((flags & FI_INJECT) && !(flags & FI_COMPLETION)) {
+			size_t total_len = pkt_size + data_len;
+			if (total_len <= rxm_ep->msg_info->tx_attr->inject_size) {
+				ret = rxm_ep_transmit_generic_lightweight(
+						rxm_ep, rxm_conn, data_len, data,
+						flags, tag, op, &tx_buf);
+				if (OFI_UNLIKELY(ret))
+					return ret;
+				ofi_copy_from_iov(tx_buf->pkt.data, tx_buf->pkt.hdr.size,
+						  iov, count, 0);
+				ret = fi_inject(rxm_conn->msg_ep, &tx_buf->pkt, total_len, 0);
+				if (OFI_UNLIKELY(ret))
+					FI_DBG(&rxm_prov, FI_LOG_EP_DATA,
+					       "fi_inject for MSG provider failed\n");
+				/* release allocated buffer for further reuse */
+				goto done_inject;
 			}
-			ret = fi_inject(rxm_conn->msg_ep, pkt, pkt_size, 0);
-			if (ret)
-				FI_DBG(&rxm_prov, FI_LOG_EP_DATA,
-				       "fi_inject for MSG provider failed\n");
-			/* release allocated buffer for further reuse */
-			goto done;
-		} else {
-			FI_DBG(&rxm_prov, FI_LOG_EP_DATA, "passed data (size = %zu) is too "
-			       "big for MSG provider (max inject size = %zd)\n",
+			FI_DBG(&rxm_prov, FI_LOG_EP_DATA, "passed data (size = %zu) "
+			       "is too big for MSG provider (max inject size = %zd)\n",
 			       pkt_size, rxm_ep->msg_info->tx_attr->inject_size);
 		}
+		ret = rxm_ep_transmit_generic(rxm_ep, rxm_conn, context, (uint8_t)count,
+					      data_len, data, flags, tag, op, comp_flags,
+					      &tx_buf, &tx_entry);
+		if (OFI_UNLIKELY(ret))
+			return ret;
+		tx_entry->state = RXM_TX;
+		pkt_size += tx_buf->pkt.hdr.size;
 	}
 
-	ret = fi_send(rxm_conn->msg_ep, pkt, pkt_size, tx_buf->hdr.desc, 0, tx_entry);
-	if (ret) {
+	ret = fi_send(rxm_conn->msg_ep, &tx_buf->pkt, pkt_size,
+		      tx_buf->hdr.desc, 0, tx_entry);
+	if (OFI_UNLIKELY(ret)) {
 		if (ret == -FI_EAGAIN)
 			rxm_cq_progress(rxm_ep);
 		else
 			FI_WARN(&rxm_prov, FI_LOG_EP_DATA,
 				"fi_send for MSG provider failed\n");
-		goto done;
+		goto done_send_mr;
 	}
 	return 0;
-done:
-	rxm_buf_release(&rxm_ep->tx_pool, (struct rxm_buf *)tx_buf);
+done_send_mr:
+	if (!rxm_ep->rxm_mr_local &&
+	    (data_len > rxm_ep->rxm_info->tx_attr->inject_size))
+		rxm_ep_msg_mr_closev(tx_entry->mr, tx_entry->count);
+done_send:
 	rxm_tx_entry_release(&rxm_ep->send_queue, tx_entry);
+done_inject:
+	rxm_buf_release(&rxm_ep->tx_pool, (struct rxm_buf *)tx_buf);
 	return ret;
 }
 
 #define rxm_ep_tx_flags_inject(ep_fid) \
 	((rxm_ep_tx_flags(ep_fid) & ~FI_COMPLETION) | FI_INJECT)
 
+#define rxm_inject(...) \
+	rxm_ep_inject_common(__VA_ARGS__, 0, ofi_op_msg, FI_MSG)
+
+#define rxm_tinject(...) \
+	rxm_ep_inject_common(__VA_ARGS__, ofi_op_tagged, FI_TAGGED)
+
 #define rxm_send(...) \
 	rxm_ep_send_common(__VA_ARGS__, 0, ofi_op_msg, FI_MSG)
```

Comment:
I'd be in favor of dropping these defines and calling the actual function directly.  The only purpose these defines serve is to save typing at the cost of making a reader need to hop through a level of coding indirection to figure out what's happening.